### ISSUE TITLE
Arabic language editing problem

### ISSUE THREAD

**Issue by AliBorsan** [notepad-plus-plus/notepad-plus-plus#10664]
There is a serious problem in Notpad++ when you try to edit an arabic text:
1 - When you try to modify a word or text by clicking before or after any character and delete or erase it, it delete another char..
2- When you select any part of any text, and replace it with something from the clipboard, it replaces another chars or words.
3- When you select a word and paste it >> you find that it copied another word.
That make editing any text or code that contain Arabic strings something very hard and sometimes impossible.
4- Text align to right is not working.
So it is not any kind of prefered editor by any Arabic coder ,translator or editor.
It is my prefered editor for English, German and Turkish, but when it comes to Arabic.. i have to use another editor.
Hope you find a fix if you care.
All the best.

**Comment by beh-mard** [comment]
I am facing exactly the same problems, it is very interesting that after about 3 years of this report, the problem has not been solved yet!

**Comment by xomx** [comment]
Yes, it does not work, some sample text to try: 
[arabic-sample-text.txt](https://github.com/user-attachments/files/16103303/arabic-sample-text.txt)

The N++ underlying editor component Scintilla has currently only experimental support for the bidi R2L langs (Arabic, Hebrew, ...):
https://scintilla.org/ScintillaDoc.html#SCI_SETBIDIRECTIONAL
More info about the Scintilla R2L implementation:
https://groups.google.com/g/scintilla-interest/c/6u5YFzkRzRE/m/hdW8WY2XAQAJ

It seem to be implemented only for the DirectWrite Scintilla mode and only for the UTF-8 document encoding and only in SC_BIDIRECTIONAL_L2R (the SC_BIDIRECTIONAL_R2L seems not to be finished yet). But when one wants to use the currently implemented N++ RTL (menu > View > Text Direction RTL or Ctrl+Alt+R), that DirectWrite mode cannot be used in N++ (menu > Settings > Preferences > MISC. > Use DirectWrite checkbox has to be off):
https://sourceforge.net/p/scintilla/bugs/2233/
`Notepad++ is using WS_EX_LAYOUTRTL. I tried using that myself and discovered that WS_EX_LAYOUTRTL only works if SC_TECHNOLOGY_DEFAULT is set. If we are using DirectWrite, it seems that we would need to call [SetReadingDirection](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritetextformat-setreadingdirection) with parameter DWRITE_READING_DIRECTION_RIGHT_TO_LEFT.`

> it is very interesting that after about 3 years of this report, the problem has not been solved yet!

Not interesting, it is completely understandable. It is certainly not an easy thing to solve that in N++, and it would also be more than appropriate to have someone here, who knows such languages, to work on it...

**Comment by beh-mard** [comment]
Thank you for your answer, yes, the problem is solved with this method, but the window is right-aligned, which is not pleasant and professional for a programmer, and the writings must be left-aligned, I hope you understand!

**Comment by AliBorsan** [comment]
Thank you for participating and sharing your experiences and ideas.
But how many N++ R2L users can make their own  copy of Scintilla and patch it, which will be lost within any update?!
As the core based on Scintilla and Scintilla has only limited support for R2L , and all tries to batch .cxx files give no final solution.
As you mentioned , N++ will be and stay L2R editor with limited R2L support, so we would need a plugin to do the job, hoping that the Uniface team will find a solution in the feature.
and we must stuck on Win32 if we need to use N++ for better R2L support.
I will try to contact " Raghda Morsy " from Unifce to find if they yet had found any solution.
Thank you.

**Comment by xomx** [comment]
@beh-mard 
> , I hope you understand!

No, I do not understand. That is exactly the problem I was trying to point out above - you need to find someone here capable of both C++ AND ARABIC. Someone who knows exactly how it should look and work for the R2L lang and is willing to devote their free time to implement it here...

**Comment by beh-mard** [comment]
@xomx 
The issue is very simple, to be able to edit Arabic or Persian characters on the L2R!

**Comment by AliBorsan** [comment]
Okay Guys! This is a comparison of the same attempt in between NP++ latest version and Win notepad:
In Notepad++  > line 1 : I had intentionally omitted the character that marked red until I finished writing the line, then i went back and placed the cursor on the place of the missing char, when i type it, it jumps lot of bits and type it directly after the next to words. the same case with delete or backspace.
But in windows notepad it acts very normal as you see in the pictures.
In Notepad++ Arabic chars can  be displayed correctly only with UTF-8 , Which the Arabic character sets never works and destroy the whole text.
I made this demonstration with one language only and one direction, to avoid any confusion with bi-lingual and bi-directions cases.
Patching the .cxx files never succeeded to override this problem.

Thank you for your participating and investigating in the problem.
All the best

![npcompare](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/80667402/5c53b2c4-3f7b-44aa-b4dc-e6be3be9b953)

**Comment by AliBorsan** [comment]
Also here is the text of you dont have or know arabic to try it your way:

اللغة العربي في برنامج محرر النصوص
اللغة العربية في برنامج محرر النصوص
Just try to copy the missing char from the anywhere if the text and paste it in the right place.

**Comment by beh-mard** [comment]
@xomx 
Just copy one of the Arabic or Persian characters and paste it in the next line! Now you can see that what is placed is not the same character that you chose!

**Comment by xomx** [comment]
Related: #11658 or https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11476#issuecomment-1094276395

I only guess here but the problem is probably connected to coordinate transformations needed for the R2L langs. Simply saying - the underlying Scintilla editor component either receives wrong position info or miscalculates that somehow for the R2L. There is a saved copy of the old MS [article](https://web.archive.org/web/20141016210815/http://msdn.microsoft.com/en-us/goglobal/bb688119) about the GDI mirroring awareness:
![coordinate-transformations](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/13075183/43523251-176a-4e72-93be-a0ebae3f0c82)

The standard Windows Notepad app seems not have this problem. Its edit-control used there for the text manipulation does not use any of the extended windows styles like WS_EX_RTLREADING (I "spied" at runtime). So it is probably a direct property of the standard Windows edit-control component. That correct "Notepad like" behavior can also be tested in the N++ where this edit-ctrl is used, e.g.:
![npp-r2l-notepad](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/13075183/7c53d6b0-1b66-4c31-8403-9546aa75c8c7)

IMO - until the Scintilla does not have better support for the R2L, there is not much to be done in N++ to fix these kind of issues. Evidently other Scintilla based editors have similar problems (search for the Geany or Notepad2 R2L issues).

@Snabel42 
> Notepad handled most edits as I'd expect, but .. keyboard keys rightarrow & leftarrow did move in the opposite direction
> Maybe that's an intentional choice?

Yes.

> But whether the text is R2L or L2R, I suppose I would have thought htat left was still left, right was still right.

No.

> Notepad++ handled right & left arrow like "normal"

In LTR. Try RTL (menu > View > Text Direction RTL or Ctrl+Alt+R).

**Comment by AliBorsan** [comment]
@xomx 
With Ctrl+Alt+R there is no problem at all which is okay for typing a normal text.
But if you are writing a code then you have no other choice than using Ctrl+Alt+L or LTR , in cases like typing messages, exception , error messages, notifications or translations into a RTL language, or editing or editing codes that has RTL strings, the thing will go very messy. 
Patching Scintilla cxx files will not solve the problem because it is not a C++ problem, it is about functions who involves some standards and default sizes of deferent chars from several languages. 
As a conclusion, I think it is not an ideal option anymore to try to use NP++ as an IDE if any RTL language could be involved. 
You did a great job @xomx , also @beh-mard and @Snabel42.
Thank you very much.

**Comment by AliBorsan** [comment]
Also When you switch the direction with  Ctrl+Alt+R , the problem leaves  the RTL text and moves to the LTR text which will have the same issue that was with his RTL brother.
So the problem is shifting with the direction to the opposite language.
For a test : in LTR mode type some text , then  Ctrl+Alt+R change the direction.
then try to insert, type, copy or delete any character in any location or position.
You will find it jumping many words after the intended position.
I think it is designed as one direction for each language family and not intended for bi-directional purposes.
It is a train case I think.
Notice the red dots in the picture which represent  Space and Tab symbols.
![np22](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/80667402/9956b488-7341-4cc6-b100-cb3c7d1de0e5)
